quinn-proto/src/new_token_store.rs

```diff
@@ -0,0 +1,228 @@
+//! Storing tokens sent from servers in NEW_TOKEN frames and using them in subsequent connections
+
+use bytes::Bytes;
+use slab::Slab;
+use std::{
+    collections::{hash_map, HashMap},
+    mem::take,
+    rc::Rc,
+    sync::Mutex,
+};
+
+/// Responsible for storing tokens sent from servers in NEW_TOKEN frames and retreiving them for
+/// use in subsequent connections
+pub trait NewTokenStore: Send + Sync {
+    /// Called when a NEW_TOKEN frame is received from the server
+    fn store(&self, server_name: &str, token: Bytes);
+
+    /// Called when trying to connect to a server
+    ///
+    /// The same token should never be returned from `take` twice.
+    fn take(&self, server_name: &str) -> Option<Bytes>;
+}
+
+/// `NewTokenStore` implementation that stores up to `N` tokens per server name for up to a limited
+/// number of server names, in-memory
+pub struct InMemNewTokenStore<const N: usize>(Mutex<InMemNewTokenStoreState<N>>);
+
+impl<const N: usize> InMemNewTokenStore<N> {
+    /// Construct empty
+    pub fn new(size_limit: usize) -> Self {
+        Self(Mutex::new(InMemNewTokenStoreState::new(size_limit)))
+    }
+}
+
+impl<const N: usize> NewTokenStore for InMemNewTokenStore<N> {
+    fn store(&self, server_name: &str, token: Bytes) {
+        self.0.lock().unwrap().store(server_name, token)
+    }
+
+    fn take(&self, server_name: &str) -> Option<Bytes> {
+        self.0.lock().unwrap().take(server_name)
+    }
+}
+
+/// Defaults to a size limit of 256
+impl<const N: usize> Default for InMemNewTokenStore<N> {
+    fn default() -> Self {
+        Self::new(256)
+    }
+}
+
+// safety: InMemNewTokenStoreState is !Send and !Sync because it uses Rc. so long as
+//         InMemNewTokenStore doesn't keep Rc handles after the mutex is unlocked, it's safe.
+unsafe impl<const N: usize> Send for InMemNewTokenStore<N> {}
+unsafe impl<const N: usize> Sync for InMemNewTokenStore<N> {}
+
+#[derive(Debug)]
+struct InMemNewTokenStoreState<const N: usize> {
```

Comment:
I don't think so. I tried rewriting this in indexmap but it didn't simplify it much. The real difficulty isn't creating an associative mapping with stable fast indices, it's creating the linked LRU cache. Maybe I missed something basic but I don't see a way to rearrange the indexes of indexmap in a way that would achieve that. I also tried pulling in a couple other Lru cache crates but none of them seemed great and I had to sacrifice some memory optimizations and I don't think it's overall worth the complexity.

That said, I have:

- Split this module up into various separate layers of data structure rather than one flattened all together, which should make it seem more tractable
- Written some very robust tests, which are repeated deterministic stochastic tests that assert that these data structures not only maintain equivalent external behavior to simpler and worse time complexity alternative implementations, but also that they maintain isomorphic internal structure as well

Because of these things, especially the point about the tests I've written, I feel pretty good about these data structures at this point.

Reply:
Splitting off into follow-up PR